resources/type-system/inference.md

```diff
@@ -1210,6 +1247,674 @@ of steps:
+static type analysis. _For example, `a.b()` could be a static method invocation
+(if `a` refers to a class name and `b` is a static method in that class), an
+instance method invocation (if `a` is an expression whose type contains an
+instance method called `b`), or a function invocation applied to the result of
+an instance get (if `a` is an expression whose type contains a getter called
+`b`), among other things._ Accordingly, the process of disambiguating and type
+inferring a selector chain requires interleaving the rules for type inference
+with the name resolution process.
+
+This is accomplished by pattern matching the selector chain against each of the
+rules below in turn. The first matching rule is chosen, and is used to type
+infer the selector chain.
+
+Some of the rules below match the entire selector chain. Other rules match a
+specific sequence of selectors on the right, in which case the remainder of the
+selector chain (if present) is then considered to be a subexpression. _So, for
+example, the selector chain `1.isEven.toString()` is considered by the grammar
+to consist of the primary `1` followed by the three selectors `.isEven`,
+`.toString`, and `()`. This matches the [method invocation](#Method-invocation)
+rule, which then treats `1.isEven` as the remainder subexpression, `toString` as
+the method name identifier, and `()` as the &lt;argumentPart&gt;._
+
+The selector chain type inference rules are as follows.
+
+### Static method invocation
+
+If the selector chain is a sequence of 1 to 3 _&lt;identifier&gt;s_ separated by
+`.`, followed by an _&lt;argumentPart&gt;_, and the sequence of
+_&lt;identifier&gt;s_ can be resolved to a static method or top level function,
+then the result of selector chain type inference in context `K` is the
+elaborated expression `m`, with static type `T`, where `m` and `T` are
+determined as follows:
+
+- Let `f` be the static method or top level function referred to by the
+  _&lt;identifier&gt;_ sequence.
+
+- Let `F` be the type of `f`.
+
+- Invoke [argument part inference](#Argument-part-inference) on
+  _&lt;argumentPart&gt;_, using `F` as the target function type and `K` as the
+  context. Designate the result by `{m_1, m_2, ...}`, `{U_1, U_2, ...}`, and
+  `R`.
+
+- Let `m` be `@STATIC_INVOKE(f<U_1, U_2, ...>(n_1: m_1, n_2: m_2, ...))`, and
+  let `T` be `R`. _This is sound because `R` is the result of substituting
+  `<U_1, U_2, ...>` for the type arguments of `f` in the return type of `f`._
+
+### Implicit instance creation
+
+If the selector chain consists of _&lt;typeName&gt; &lt;typeArguments&gt;_? (`.`
+_&lt;identifierOrNew&gt;_)? _&lt;arguments&gt;_, and _&lt;typeName&gt;_ can be
+resolved to a type in the program, then the result of selector chain type
+inference in context `K` is the elaborated expression `m`, with static type `T`,
+where `m` and `T` are determined as follows:
+
+_TODO(paulberry): specify this._
+
+### Implicit this invocation
```

Comment:
I may be missing something, but this seems wrong.  If I follow the algorithm here, I think it implies that the following is an error:
```
int Function(int) f = (x) => x;
void test() {
 f(3);
}
```

because:
  - Neither of the previous two cases apply to `f(3)`
  -  I think `f(3)` parses so as to match this case
  - The selector chain here does not have access to `this`.

Am I missing something?

Reply:
You’re right. As a band-aid, I’ve changed the text “__cannot__ be resolved to the name of a local variable” to “__cannot__ be resolved using local scope resolution rules” to resolve this problem, but this is not a good solution. In particular, it doesn’t handle the possibility that the identifier _can_ be resolved using local scope resolution rules, and it resolves to an instance method or getter in the current class.

I’ve been working ahead a bit while this PR was under review, and I’ve got some ideas for how to address this, so I’m going to defer further work on this issue to a TODO comment.